src/lib/operators/jit_operator/operators/jit_aggregate.cpp

```diff
@@ -0,0 +1,312 @@
+#include "jit_aggregate.hpp"
+
+#include "constant_mappings.hpp"
+#include "operators/jit_operator/jit_operations.hpp"
+#include "resolve_type.hpp"
+#include "storage/value_column.hpp"
+
+namespace opossum {
+
+std::string JitAggregate::description() const {
+  std::stringstream desc;
+  desc << "[Aggregate] GroupBy: ";
+  for (const auto& groupby_column : _groupby_columns) {
+    desc << groupby_column.column_name << " = x" << groupby_column.tuple_value.tuple_index() << ", ";
+  }
+  desc << " Aggregates: ";
+  for (const auto& aggregate_column : _aggregate_columns) {
+    desc << aggregate_column.column_name << " = " << aggregate_function_to_string.left.at(aggregate_column.function)
+         << "(x" << aggregate_column.tuple_value.tuple_index() << "), ";
+  }
+  return desc.str();
+}
+
+std::shared_ptr<Table> JitAggregate::create_output_table(const uint32_t max_chunk_size) const {
+  auto num_columns = _aggregate_columns.size() + _groupby_columns.size();
+  TableColumnDefinitions column_definitions(num_columns);
+
+  // Add each groupby column to the output table
+  for (const auto& column : _groupby_columns) {
+    const auto data_type = column.hashmap_value.data_type();
+    const auto is_nullable = column.hashmap_value.is_nullable();
+    column_definitions[column.table_position] = {column.column_name, data_type, is_nullable};
+  }
+
+  // Add each aggregate to the output table
+  for (const auto& column : _aggregate_columns) {
+    // Computed averages are always of type double. The hashmap_value, however, is used to compute a SUM aggregate
+    // (that is lated used to produce averages in a post-processing step) and may thus have a different data type.
+    const auto data_type =
+        column.function == AggregateFunction::Avg ? DataType::Double : column.hashmap_value.data_type();
+    const auto is_nullable = column.hashmap_value.is_nullable();
+    column_definitions[column.table_position] = {column.column_name, data_type, is_nullable};
+  }
+
+  return std::make_shared<Table>(column_definitions, TableType::Data, Chunk::MAX_SIZE);
+}
+
+void JitAggregate::before_query(Table& out_table, JitRuntimeContext& context) const {
+  // Resize the hashmap data structure.
+  context.hashmap.values.resize(_num_hashmap_values);
```

Comment:
How can we know the number of values in the hashmap before we even started the query? Naming issue?

Reply:
Renamed to `columns`.